File: 2069-walking-robot-simulation-ii/2069-walking-robot-simulation-ii.java
class Robot {
    
    int arr[][];
    int x=0,y=0,h=0,w=0;
    String dir="East";
    public Robot(int width, int height) {
        arr=new int[height][width];
        h=height;
        w=width;
        // System.out.println(width+" "+height);
    }
    
    public void step(int num) {
        
        long temp=(h-1)*2 +(w-1)*2;
        num=(int)((long)num%(temp));
        if (x == 0 && y == 0){
            dir="South";
        }
        while(num-->0){
            if(dir.equals("East")){
                if(y==w-1){
                    x++;
                    dir="North";
                }
                else y++;
            }
            else if(dir.equals("North")){
                if(x==h-1){
                    y--;
                    dir="West";
                }
                else x++;
            }
            else if(dir.equals("West")){
                if(y==0){
                    x--;
                    dir="South";
                }
                else y--;
            }
            else if(dir.equals("South")){
                if(x==0){
                    y++;
                    dir="East";
                }
                else x--;
            }
        }
        
    }
    
    public int[] getPos() {
        return new int[]{y,x};
    }
    
    public String getDir() {
        return dir;
    }
}

/**
 * Your Robot object will be instantiated and called as such:
 * Robot obj = new Robot(width, height);
 * obj.step(num);
 * int[] param_2 = obj.getPos();
 * String param_3 = obj.getDir();
 */

File: 2069-walking-robot-simulation-ii/2069-walking-robot-simulation-ii.cpp
class Robot {
    int dir = 0;
    int m, n;
    int i, j;
public:
    Robot(int width, int height) {
        i = 0;
        j = 0;
        m = height;
        n = width;
    }
    
    void step(int num) {
        num = num % (2 * (n - 1) + 2 * (m - 1));
        if (i == 0 && j == 0) dir = 3; // this line was missing from contest submission
        
        while (num > 0) {
            if (dir == 0) {
                if (j + num < n) {
                    j += num;
                    num = 0;
                } else {
                    dir = 1;
                    num -= n - 1 - j;
                    j = n - 1;
                }
            } else if (dir == 1) {
                if (i + num < m) {
                    i += num;
                    num = 0;
                } else {
                    dir = 2;
                    num -= m - 1 - i;
                    i = m - 1;
                }
            } else if (dir == 2) {
                if (j >= num) {
                    j -= num;
                    num = 0;
                } else {
                    num -= j;
                    j = 0;
                    dir = 3;
                }
            } else {
                if (i >= num) {
                    i -= num;
                    num = 0;
                } else {
                    dir = 0;
                    num -= i;
                    i = 0;
                }
            }
        }
    }
    
    vector<int> getPos() {
        return vector<int>{j, i};
    }
    
    string getDir() {
        if (dir == 0) return "East";
        if (dir == 1) return "North";
        if (dir == 2) return "West";
        return "South";
    }
};

/**
 * Your Robot object will be instantiated and called as such:
 * Robot* obj = new Robot(width, height);
 * obj->step(num);
 * vector<int> param_2 = obj->getPos();
 * string param_3 = obj->getDir();
 */